scheduler/planner.go

```diff
@@ -195,78 +196,93 @@ type unitInfo struct {
 	ContainsStepbackTask bool `json:"contains_stepback_task"`
 }
 
-func (u *unitInfo) value() int64 {
-	var value int64
-
-	length := int64(len(u.TaskIDs))
-	priority := 1 + (u.TotalPriority / length)
-
-	if !u.ContainsNonGroupTasks {
-		// if all tasks in the unit are in a task group then
-		// we should give it a little bump, so that task
-		// groups tasks are sorted together even when they
-		// would also be scheduled in a version.
-		priority += length
-	}
-	if u.ContainsGenerateTask {
-		// give generators a boost so people don't have to wait twice.
-		priority = priority * u.Settings.GetGenerateTaskFactor()
-	}
+// value computes a full SortingValueBreakdown, containing the final value by which the unit
+// will be sorted in its queue (stored in TotalValue), and the numerical influences that
+// the unit's properties had on computing that final value. Currently, the formula for
+// computing this value is (custom_priority * custom_rankValue) + unit_length, where custom_priority
+// and custom_rankValue are both derived from specific properties of the unit and various
+// scheduler constants.
+func (u *unitInfo) value() task.SortingValueBreakdown {
+	var breakdown task.SortingValueBreakdown
+	priority := u.computePriority(&breakdown)
+	rankValue := u.computeRankValue(breakdown.TaskGroupLength, &breakdown)
+	breakdown.TotalValue = priority*rankValue + breakdown.TaskGroupLength
+	return breakdown
+}
 
+// computeRankValue computes the custom rank value for this unit, which will later be multiplied with the
+// custom priority to compute a final value by which the unit will be sorted on in the queue.
+func (u *unitInfo) computeRankValue(unitLength int64, breakdown *task.SortingValueBreakdown) int64 {
```

Comment:
do we have to pass in unitLength separately, since it just comes from breakdown? If we just want the original unit length (like we update it at some point), we could just set like `originalUnitLength = breakdown.TaskGroupLength` at the beginning of this function

Reply:
Yes that's true. Updated